src/family_name_classifier.py

```diff
@@ -0,0 +1,315 @@
+
+        print("Generating dataset matrix.")
+        # Iterates over the pairs of names to generate the features dictionaries
+        self.vectorizer = sklearn.feature_extraction.DictVectorizer()
+        dataset = self.vectorizer.fit_transform(map(lambda tup: self.feature_dict(tup[0], tup[1]),
+                                                    names_iterator))
+        n_positives = dataset.shape[0] - n_negative_samples
+        target = [0] * n_negative_samples + [1] * n_positives
+
+        print("Training")
+        scoring = 'precision'
+        self.model = sklearn.linear_model.LogisticRegressionCV(n_jobs=-1, scoring=scoring)
+        self.model.fit(dataset, target)
+
+        # Debug part:
+        print("\nBest", scoring, "score:", np.max(np.average(self.model.scores_[1], axis=0)))
+        importance_order = np.argsort(self.model.coef_[0, :])
+        feat_names = self.vectorizer.get_feature_names()
+        print("Trained bias weight:", self.model.intercept_[0])
+        print("Trained weights of features:")
+        for idx_feat in list(importance_order):
+            print(feat_names[idx_feat], self.model.coef_[0, idx_feat])
+
+        if file_name_save is not None:
+            print("\nSaving...")
+            self.save(file_name_save)
+
+    def family_probability(self, name1, name2):
+        """
+        Using the trained model, gives a probability that two names belongs to the same family.
+
+        :param name1: string
+        :param name2: string
+        :return:
+        """
+        x = self.vectorizer.transform(self.feature_dict(name1, name2))
+        probabilities = self.model.predict_proba(x)
+        return probabilities[0, 1]
+
+    def same_family(self, group1, group2, threshold=0.5, return_match=False):
+        """
+        Given two family groups, it says if there is any combination of persons that
+        may be relatives, using the trained algorithm and a probability threshold to decide.
+
+        If return_match is true, this returns a tuple of the first pair of probable relatives found.
+
+        :param group1: list of string
+        :param group2: list of string
+        :param threshold: float
+        :param return_match: boolean
+        :return: boolean or tuple
+        """
+        for person1, person2 in itertools.product(group1, group2):
+            prob_relatives = self.family_probability(person1, person2)
+            if prob_relatives >= threshold:
+                if return_match:
+                    return person1, person2
+                return True
+
+        if return_match:
+            return None
+        return False
+
+    def save(self, file_name):
+        """
+        Saves the current (trained) object with scikit-learn Joblib method.
+
+        :param file_name: string
+        :return:
+        """
+        sklearn.externals.joblib.dump(self, file_name, compress=9)
+
+    @builtins.staticmethod
```

Comment:
Could be an external function too: it is really just to simplify the load of a saved classifier.